okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.java

```diff
@@ -131,9 +139,27 @@ private void readHeader() throws IOException {
     boolean reservedFlag1 = (b0 & B0_FLAG_RSV1) != 0;
     boolean reservedFlag2 = (b0 & B0_FLAG_RSV2) != 0;
     boolean reservedFlag3 = (b0 & B0_FLAG_RSV3) != 0;
-    if (reservedFlag1 || reservedFlag2 || reservedFlag3) {
+
+    if (reservedFlag1 && !compressionEnabled) {
+      throw new ProtocolException(
+          "Reserved flag rsv1 is not supported if compression is disabled.");
+    }
+
+    if (reservedFlag2 || reservedFlag3) {
       // Reserved flags are for extensions which we currently do not support.
-      throw new ProtocolException("Reserved flags are unsupported.");
+      throw new ProtocolException("Reserved flags rsv2 and rsv3 are unsupported.");
+    }
+
+    if (reservedFlag1 && isControlFrame) {
+      throw new ProtocolException("Reserved flag rsv1 must not be set on control frames.");
+    }
+
+    if (reservedFlag1 && opcode == OPCODE_CONTINUATION) {
```

Comment:
can you refactor this so we read reservedFlag1 only once?

Reply:
Indeed :) Done.